add more parameters to custom retry policy

timbussmann opened the pull request:
fixes https://github.com/Particular/NServiceBus/issues/3831 by adding additional overloads for the custom retry policy which also grants access to the causing exception and the retry attempt.

There are now 3 overloads available:
* `Func<IncomingMessage, TimeSpan> customPolicy`
* `Func<IncomingMessage, Exception, TimeSpan> customPolicy`
* `Func<IncomingMessage, Exception, int, TimeSpan> customPolicy`

I made one only using the exception with the int parameter because I think the int value not very intuitive (should probably a custom type called `SlrRetryAttempt` to make more sense).

Also fixed the acceptance test which missed that bug all the time.

@Particular/nservicebus-maintainers @mauroservienti please review

docs PR: https://github.com/Particular/docs.particular.net/pull/1602

timbussmann commented:
based on a discussion with @bording an alternative to all these overloads would be to use a single parameter like `ProcessingFailureContext` with the three parameters as properties on it (so we can name the int properly).

timbussmann commented:
changed the API to be `Func<SecondLevelRetryContext, TimeSpan> customPolicy` instead of the overloads. Ready much better :+1:

Scooletz commented:
Would you consider providing a custom delegate for this rather than a specified `Func`. If you write an xml comment for the delegate it will be accessible for consumers even if they pass it elsewhere. It does not work for `Func`.

timbussmann commented:
@Scooletz In general that sounds like a reasonable idea, but I remember others having this conversation already a few times (I even think we reverted some newly introduced delegates).

I know @SimonCropp having some valid arguments:
> the problem with the custom delegate is u can never evolve it forward in a non breaking way. u need to add a new "2" version. the func approach u can add a new config overload with a new func definition

Since this API already was using funcs and the new parameter allows us to describe&name the parameters better already, I think I'd stay with the current proposal for now.

bording commented on src/NServiceBus.Core/Recoverability/SecondLevelRetries/CustomSecondLevelRetryPolicy.cs:
```diff
@@ -1,22 +1,20 @@
         }
 
-        public override bool TryGetDelay(IncomingMessage message, Exception ex, int currentRetry, out TimeSpan delay)
+        Func<SecondLevelRetryContext, TimeSpan> customRetryPolicy;
```
Is this the right spot for this?

bording commented on src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetryContext.cs:
```diff
@@ -0,0 +1,26 @@
+        public IncomingMessage Message { get; set; }
+
+        /// <summary>
+        /// The occured exception.
```
"The exception that occurred." sounds better.

bording commented on src/NServiceBus.Core/Recoverability/SecondLevelRetries/Config/SecondLevelRetriesSettings.cs:
```diff
@@ -19,16 +18,17 @@ internal SecondLevelRetriesSettings(EndpointConfiguration config)
 
         /// <summary>
-        /// Register a custom retry policy.
+        /// Register a custom retry policy. The callback receives the failed message, the exception and the current second level retry attempt.
```
Registers. Also, I prefer a comma after "exception".

bording commented on src/NServiceBus.Core/Recoverability/SecondLevelRetries/Config/SecondLevelRetriesSettings.cs:
```diff
@@ -19,16 +18,17 @@ internal SecondLevelRetriesSettings(EndpointConfiguration config)
+        /// Register a custom retry policy. The callback receives the failed message, the exception and the current second level retry attempt.
         /// </summary>
-        public void CustomRetryPolicy(Func<IncomingMessage, TimeSpan> customPolicy)
+        /// <param name="customPolicy">the function which is invoked on a failed message to determine the delay until the message is retried.</param>
```
Capitalize "the" and replace "which" with "that"

bording commented on src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetryContext.cs:
```diff
@@ -0,0 +1,26 @@
+    public class SecondLevelRetryContext
+    {
+        /// <summary>
+        /// The message which failed to process.
```
"that" works better here than "which"

bording commented on src/NServiceBus.AcceptanceTests/Recoverability/Retries/When_performing_slr_with_min_policy.cs:
```diff
@@ -46,7 +45,7 @@ public RetryEndpoint()
                     .WithConfig<SecondLevelRetriesConfig>(c => c.TimeIncrease = TimeSpan.FromMilliseconds(1));
```
Not part of your changes, but with a custom retry policy, is there even a point to specifying a SecondLevelRetriesConfig here? Looks like several tests do this.

timbussmann replied:
no, that setting will be overridden. Good catch! :+1:

timbussmann commented:
@bording addressed your comments :+1: 

since there where no other comments over the last days I'm going to merge this once builds are green. Feel free to continue discussions on the main PR: https://github.com/Particular/NServiceBus/pull/3814

danielmarbach commented on src/NServiceBus.AcceptanceTests/Recoverability/Retries/When_performing_slr_with_custom_policy.cs:
```diff
@@ -1,34 +1,39 @@
-            public bool FLRetriesHeaderAvailable { get; set; }
             public bool MessageMovedToErrorQueue { get; set; }
-            public bool RetriesHeaderAvailable { get; set; }
+            public List<SecondLevelRetryContext> HeaderValues { get; } = new List<SecondLevelRetryContext>();
```
Is the name of the property still accurate? Since the name `SecondLevelRetryContext` doesn't imply header values

timbussmann replied:
good catch, should be renamed

danielmarbach commented on src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt:
```diff
@@ -975,6 +975,13 @@ namespace NServiceBus
+        public SecondLevelRetryContext() { }
+        public System.Exception Exception { get; set; }
+        public NServiceBus.Transports.IncomingMessage Message { get; set; }
+        public int SecondLevelRetryAttempt { get; set; }
```
NumberOfAttemps? Not sure if repeating the word SLR provides any meaningful context

timbussmann replied:
I wanted to avoid "NumberOfAttempts" since that could also mean **total** number of retry attempts. I think it adds value.

danielmarbach replied:
The first sentence is a valid and good argument. I don't see the "adds value" part in the second sentence. How does prefixing a property with a `SecondLevelRetry` in the context of a second level retry setting under a second level retry context provide any value? I fail to see this. I'd just name this `Attempt` then

danielmarbach replied:
It seems @timbussmann has already discussed my concerns intensively with @bording I don't want to hold up this work any longer then.

danielmarbach commented on src/NServiceBus.Core.Tests/Recoverability/SecondLevelRetries/SecondLevelRetriesTests.cs:
```diff
@@ -190,7 +186,10 @@ IncomingMessage CreateMessage(string id = "id", string slrRetryHeader = null, by
             var headers = string.IsNullOrEmpty(slrRetryHeader)
                 ? new Dictionary<string, string>()
-                : new Dictionary<string, string> { { Headers.Retries, slrRetryHeader }};
+                : new Dictionary<string, string>
```
This is done by our template right? I usually prefer one liners for single entry dictionaries. But not a big deal

timbussmann replied:
yes, that's the template
